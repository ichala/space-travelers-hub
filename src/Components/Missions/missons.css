table {
  border-collapse: collapse;
  width: 100%;
}

th,
td {
  padding: 1.25rem;
  border: 1px solid #ccc;
  text-align: center;
}

.joined_mission {
  background-color: rgb(47, 176, 0);
  padding: 3px 9px 3px 9px;
  border-radius: 8px;
  color: white;
}

.not_joined_mission {
  background-color: rgb(98, 99, 98);
  padding: 3px 9px 3px 9px;
  border-radius: 8px;
  color: white;
}

.join_mission {
  background-color: transparent;
  color: black;
  padding: 5px;
  border-radius: 5px;
  border: 2px solid black; /* Green */
}

.join_mission:hover {
  background-color: black;
  color: rgb(255, 255, 255);
  padding: 5px;
  border-radius: 5px;
  border: 2px solid black;
  cursor: pointer;
}

.leave_mission {
  background-color: transparent;
  color: rgb(223, 0, 0);
  padding: 5px;
  border-radius: 5px;
  border: 2px solid rgb(227, 8, 8);
}

.leave_mission:hover {
  background-color: rgb(223, 0, 0);
  color: rgb(255, 255, 255);
  padding: 5px;
  border-radius: 5px;
  border: 2px solid rgb(227, 8, 8);
  cursor: pointer;
}

.table-description {
  width: 55%;
}

.table-title,
.table-status,
.table-action {
  width: 15%;
}

tbody tr:nth-child(odd) {
  background: #eee;
}

.table-container {
  padding: 20px;
  margin: 40px;
}
