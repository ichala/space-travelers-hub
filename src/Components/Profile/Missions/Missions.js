import React from 'react';
import { useSelector } from 'react-redux';

function Missions() {
  const missions = useSelector((state) => state.missions);
  const filteredMissions = missions.filter((mission) => mission.reserved);
  return (
    <table className="zebra">
      <thead>
        <tr>
          <th className="table-title">Missions</th>
        </tr>
      </thead>
      <tbody>
        {filteredMissions.length < 1 && (
        <tr>
          <td className="table-title">No missions available ..</td>
        </tr>
        )}
        {filteredMissions.map((mission) => (
          <div key={mission.name}>
            <tr>
              <td className="table-title">{mission.name}</td>
            </tr>
          </div>
        ))}
      </tbody>
    </table>
  );
}

export default Missions;
