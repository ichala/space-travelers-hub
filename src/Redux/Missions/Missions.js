const GET = 'GET_MISSIONS';
const JOIN = 'JOIN_MISSION';
const LEAVE = 'LEAVE_MISSION';
const MissionAPI = 'https://api.spacexdata.com/v3/missions';
let newState = [];
// Api Calls
export const FetchMessions = async () => {
  const response = await fetch(MissionAPI);
  const data = await response.json();
  const resultData = [];
  data.forEach((item) => {
    resultData.push({
      id: item.mission_id,
      name: item.mission_name,
      description: item.description,
      reserved: false,
    });
  });
  return resultData;
};

// Reducer
export default function reducerMissions(state = [], action = {}) {
  switch (action.type) {
    case GET:
      return action.payload;
    case JOIN:
      newState = state.map((mission) => {
        if (mission.id !== action.payload) return mission;
        return { ...mission, reserved: true };
      });
      return newState;
    case LEAVE:
      newState = state.map((mission) => {
        if (mission.id !== action.payload) return mission;
        return { ...mission, reserved: false };
      });
      return newState;
    default:
      return state;
  }
}

// Action Creators
export function SetMissions(missions) {
  return { type: GET, payload: missions };
}

export function JoinMission(id) {
  return { type: JOIN, payload: id };
}

export function LeaveMission(id) {
  return { type: LEAVE, payload: id };
}

export const MissionList = (state) => state.missions;
