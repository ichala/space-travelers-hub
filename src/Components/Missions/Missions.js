import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { JoinMission, LeaveMission } from '../../Redux/Missions/Missions';
import './missons.css';

function Missions() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions);
  return (
    <>
      <div className="table-container">
        <table className="zebra">
          <thead>
            <tr>
              <th className="table-title">Missions</th>
              <th className="table-description">Description</th>
              <th className="table-status">Status</th>
              <th className="table-action">Action</th>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (

              <tr key={mission.name}>
                <td className="table-title">{mission.name}</td>
                <td className="table-description">
                  {mission.description}
                </td>
                <td className="table-status">
                  {mission.reserved ? (<kbd className="joined_mission">Active Member</kbd>) : (<kbd className="not_joined_mission">Not A Member</kbd>)}
                </td>
                <td className="table-action">
                  {mission.reserved ? (<button onClick={() => dispatch(LeaveMission(mission.id))} className="leave_mission" type="button">Leave Mission</button>) : (<button onClick={() => dispatch(JoinMission(mission.id))} className="join_mission" type="button">Join Mission</button>)}

                </td>
              </tr>

            ))}
          </tbody>
        </table>
      </div>
    </>
  );
}

export default Missions;
