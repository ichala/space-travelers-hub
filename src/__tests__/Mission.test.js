import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Missions from '../Components/Missions/Missions';
import store from '../Redux/configureStore';

it('Missions Components Renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}><Missions /></Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
