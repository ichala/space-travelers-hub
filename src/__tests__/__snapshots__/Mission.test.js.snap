// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Components Renders correctly 1`] = `
<div
  className="table-container"
>
  <table
    className="zebra"
  >
    <thead>
      <tr>
        <th
          className="table-title"
        >
          Missions
        </th>
        <th
          className="table-description"
        >
          Description
        </th>
        <th
          className="table-status"
        >
          Status
        </th>
        <th
          className="table-action"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
